typescript/infra/scripts/warp-routes/monitor/monitor-warp-routes-balances.ts

```diff
@@ -0,0 +1,239 @@
+import { PopulatedTransaction } from 'ethers';
+
+import {
+  ChainMap,
+  ChainMetadata,
+  CoinGeckoTokenPriceGetter,
+  EvmHypXERC20Adapter,
+  EvmHypXERC20LockboxAdapter,
+  IHypXERC20Adapter,
+  MultiProtocolProvider,
+  Token,
+  TokenStandard,
+  WarpCore,
+} from '@hyperlane-xyz/sdk';
+import { ProtocolType, objMerge } from '@hyperlane-xyz/utils';
+
+import { getWarpCoreConfig } from '../../../config/registry.js';
+import {
+  DeployEnvironment,
+  getRouterConfigsForAllVms,
+} from '../../../src/config/environment.js';
+import { fetchGCPSecret } from '../../../src/utils/gcloud.js';
+import { startMetricsServer } from '../../../src/utils/metrics.js';
+import { getArgs, withWarpRouteIdRequired } from '../../agent-utils.js';
+import { getEnvironmentConfig } from '../../core-utils.js';
+
+import {
+  metricsRegister,
+  updateTokenBalanceMetrics,
+  updateXERC20LimitsMetrics,
+} from './metrics.js';
+import { WarpRouteBalance, XERC20Limit } from './types.js';
+import { gracefullyHandleError, logger } from './utils.js';
+
+async function main() {
+  const { checkFrequency, environment, warpRouteId } =
+    await withWarpRouteIdRequired(getArgs())
+      .describe('checkFrequency', 'frequency to check balances in ms')
+      .demandOption('checkFrequency')
+      .alias('v', 'checkFrequency') // v as in Greek letter nu
+      .number('checkFrequency')
+      .parse();
+
+  startMetricsServer(metricsRegister);
+
+  const envConfig = getEnvironmentConfig(environment);
+  const registry = await envConfig.getRegistry();
+  const chainMetadata = await registry.getMetadata();
+
+  // The Sealevel warp adapters require the Mailbox address, so we
+  // get router configs (that include the Mailbox address) for all chains
+  // and merge them with the chain metadata.
+  const routerConfig = await getRouterConfigsForAllVms(
+    envConfig,
+    await envConfig.getMultiProvider(),
+  );
+  const multiProtocolProvider = new MultiProtocolProvider(
+    objMerge(chainMetadata, routerConfig),
```

Comment:
I wouldn't expect a router config to get merged into a chain metadata object. Maybe routerConfig is an imperfect name?

Reply:
will isolate the mailboxes to make it more clear